[AutoDiff] Change `@differentiable` syntax to use parameter names.

dan-zheng opened the pull request:
Use parameter names instead of parameter indices in `@differentiable` attribute `wrt` list.
The old syntax used parameter indices with a prefix dot, e.g. `wrt: (.0, .1)`.
The new syntax is more readable and understandable when good parameter names are chosen.

New syntax:
```swift
@differentiable(wrt: (self, input))
func applied(to input: Float) -> Output { ... }
```

The internal, lowered representation of parameter indices is unchanged. This patch affects
only AST attribute parsing/print.
Update `wrt` clause usages in stdlib and tests.

Todos:
- Support parameter name `self` in `wrt` clause, probably via escape backticks: `` `self` ``.
- Low priority: add fixits/specific diagnostics for old parameter index style.

dan-zheng committed:
[AutoDiff] Change `@differentiable` syntax to use parameter names.

Use parameter names instead of parameter indices in `@differentiable` attribute `wrt` list.
The old syntax used parameter indices with a prefix dot, e.g. `wrt: (.0, .1)`.
The new syntax is more readable and understandable when good parameter names are chosen.

New syntax:
```
@differentiable(wrt: (self, input))
func applied(to input: Float) -> Output { ... }
```

The internal, lowered representation of parameter indices is unchanged. This patch affects
only AST attribute parsing/print.
Update `wrt` clause usages in stdlib and tests.

Todos:
- Support parameter name `self` in `wrt` clause, probably via escape backticks: ``self``.
- Low priority: add fixits/specific diagnostics for old parameter index style.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented:
Great! Another change I'd like to apply is to not require parentheses when a wrt-list has one parameter. This makes the wrt-list match the syntax of the return type of the pullback.

rxwei commented on utils/gyb_syntax_support/AttributeNodes.py:
```diff
@@ -199,22 +199,11 @@
-         ]),
-
     # differentiation-func-specifier ->
     #     ('primal' | 'adjoint' | 'jvp' | 'vjp') ':' decl-name ','?
```
Apparently I forgot to remove primal/adjoint from this file. Can you help?

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22000/commits/aec0072e4eb43e7c59c6b73d980030413deaceea.

rxwei commented on test/AutoDiff/differentiable_attr_type_checking.swift:
```diff
@@ -22,7 +22,7 @@ func jvpSimpleJVP(x: Float) -> (Float, ((Float) -> Float)) {
 }
 
-@differentiable(wrt: (.1), jvp: jvpWrtSubsetJVP)
+@differentiable(wrt: (y), jvp: jvpWrtSubsetJVP)
```
It would make a lot more sense if this looked like
```swift
@differentiable(wrt: y, jvp: jvpWrtSubsetJVP)
```

dan-zheng replied:
Good point. Added support for single `wrt` parameter without parentheses in https://github.com/apple/swift/pull/22000/commits/e37fec4f5f23135145a8b4c3b25a4d15c8e1a28e.
This enables `wrt: self` for instance methods.

rxwei commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2380,8 +2384,17 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
-      case ParsedAutoDiffParameter::Kind::Index: {
-        unsigned index = parsedWrtParams[i].getIndex();
+      case ParsedAutoDiffParameter::Kind::Named: {
+        auto nameIter = std::find(originalParamNames.begin(),
```
Use `llvm::find` with `originalParamNames` as range.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22000/commits/aec0072e4eb43e7c59c6b73d980030413deaceea.

rxwei commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2260,6 +2260,10 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
       originalParams.getArray(),
       [&](ParamDecl *decl) { return decl->getInterfaceType(); });
 
+  auto originalParamNames = map<SmallVector<Identifier, 8>>(
```
This map is a bit inefficient. If all you need is to find the index, `llvm::find_if` will get you there without creating a new vector.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22000/commits/aec0072e4eb43e7c59c6b73d980030413deaceea.

rxwei commented on lib/AST/Attr.cpp:
```diff
@@ -573,23 +575,27 @@ bool DeclAttribute::printImpl(ASTPrinter &Printer, const PrintOptions &Options,
     };
 
     // Print differentiation parameters, if any.
+    auto originalParamNames = map<SmallVector<Identifier, 8>>(
```
Same here. It'd be more efficient if you just used `original->getParameters()->get(index)->getName().str()` in the loop.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22000/commits/aec0072e4eb43e7c59c6b73d980030413deaceea.

rxwei commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2380,8 +2384,17 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
+        unsigned index = std::distance(originalParamNames.begin(), nameIter);
         if ((int)index <= lastIndex) {
           TC.diagnose(paramLoc,
                       diag::differentiable_attr_wrt_indices_must_be_ascending);
```
This "must be ascending" diagnostic no longer makes sense now that there are no indices. Some rewording would be nice. Similarly, the check below for "index out of bounds" also doesn't make sense.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22000/commits/aec0072e4eb43e7c59c6b73d980030413deaceea.

rxwei commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -1466,11 +1466,11 @@ ERROR(attr_implements_expected_member_name,PointsToFirstBadToken,
 ERROR(attr_differentiable_expected_parameter,PointsToFirstBadToken,
-      "expected a parameter, which can be the index of a function parameter with a leading dot (e.g. '.0'), or 'self'", ())
+      "expected a parameter, which can be a function parameter name or 'self'", ())
 ERROR(attr_differentiable_parameter_index_must_be_positive_integer,PointsToFirstBadToken,
```
Remove this? We shouldn't talk about "index" to the user when there's no index.

dan-zheng replied:
Good point. I changed the message to `parameter names must be specified in original order` in https://github.com/apple/swift/pull/22000/commits/aec0072e4eb43e7c59c6b73d980030413deaceea.

rxwei commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -1466,11 +1466,11 @@ ERROR(attr_implements_expected_member_name,PointsToFirstBadToken,
+      "expected a parameter, which can be a function parameter name or 'self'", ())
 ERROR(attr_differentiable_parameter_index_must_be_positive_integer,PointsToFirstBadToken,
       "parameter index must be a positive integer", ())
 ERROR(attr_differentiable_nondifferentiable_function,none,
```
This and the next diagnostics seem to be obsolete and have no users. Can you help remove?

dan-zheng replied:
Thanks! Done in https://github.com/apple/swift/pull/22000/commits/aec0072e4eb43e7c59c6b73d980030413deaceea.

dan-zheng committed:
Address comments from @rxwei.

- Don't allocate auxiliary vector.
- Update parameter index related diagnostics.

dan-zheng committed:
Support single `wrt` parameter without parentheses.

Single `wrt` parameters don't require parenthesis.
The following are now valid:
```
@differentiable(wrt: x)
@differentiable(wrt: self)
```

Multiple parameters still require parentheses: `@differentiable(wrt: x, y)`.

dan-zheng commented:
Todos:
* Update stdlib to use single differentiation parameter where possible (e.g. `wrt: (self)` -> `wrt: self`).
* Update `swift/utils/gyb_syntax_support` to support single differentiation parameters (e.g. `wrt: self`) in addition to parameter lists (e.g. `wrt: (self, x, y)`). This blocks stdlib compilation.

dan-zheng committed:
Rewrite stdlib usages of `@differentiable`, update libSyntax support.

Rewrite stdlib usages of `@differentiable` with one differentiation parameter:
`wrt: (self)` -> `wrt: self`.

Note: printing still always prints parentheses around the differentiation
parameter list.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Fix `test/Syntax/round_trip_parse_gen.swift`.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented:
@swift-ci please test tensorflow
